lnrpc/rpc.proto

```diff
@@ -1618,14 +1633,104 @@ message OpenChannelRequest {
 
     /// Whether unconfirmed outputs should be used as inputs for the funding transaction.
     bool spend_unconfirmed = 12 [json_name = "spend_unconfirmed"];
+
+    /**
+    Funding shims are an optional argument that allow the caller to intercept
+    certain funding functionality. For example, a shim can be provided to use a
+    particular key for the commitment key (ideally cold) rather than use one
+    that is generated by the wallet as normal, or signal that signing will be
+    carried out in an interactive manner (PSBT based).
+    */
+    FundingShim funding_shim = 13;
 }
 message OpenStatusUpdate {
     oneof update {
         PendingUpdate chan_pending = 1 [json_name = "chan_pending"];
         ChannelOpenUpdate chan_open = 3 [json_name = "chan_open"];
+   }
+
+    /**
+    The pending channel ID of the created channel. This value may be used to
+    further the funding flow manually via the FundingStateStep method.
+    */
+    bytes pending_chan_id = 4 [json_name = "pending_chan_id"];
+}
+
+message KeyLocator {
+    /// The family of key being identified.
+    int32 key_family = 1;
+
+    /// The precise index of the key being identified.
+    int32 key_index = 2;
+}
+
+message KeyDescriptor {
+     /**
+     The raw bytes of the key being identified. Either this or the KeyLocator
+     must be specified.
+     */
+     bytes raw_key_bytes = 1;
+
+     /** 
+     The key locator that identifies which key to use for signing. Either this
+     or the raw bytes of the target key must be specified.
+     */
+     KeyLocator key_loc = 2;
+}
+
+message ChanPointShim {
+    /**
+    The size of the pre-crafted output to be used as the channel point for this
+    channel funding.
+    */
+    int64 amt = 1;
+
+    /// The target channel point to refrence in created commitment transactions.
+    ChannelPoint chan_point = 2;
+
+    /// Our local key to use when creating the multi-sig output.
+    KeyDescriptor local_key = 3;
+
+    /// The key of the remote party to use when creating the multi-sig output.
+    bytes remote_key = 4;
+
+    /**
+    If non-zero, then this will be used as the pending channel ID on the wire
+    protocol to initate the funding request. This is an optional field, and
+    should only be set if the responder is already expecting a specific pending
+    channel ID.
+    */
+    bytes pending_chan_id = 5;
+}
+
+message FundingShim {
+    oneof shim {
+        ChanPointShim chan_point_shim = 1;
     }
 }
 
+message FundingShimCancel {
+    /// The pending channel ID of the channel to cancel the funding shim for.
+    bytes pending_chan_id = 1;
+}
+
+message FundingTransitionMsg {
+    oneof trigger {
```

Comment:
What's the rational behind having a oneof here, instead of just two RPC endpoints? Seems simpler not having to deal with the oneonf, and the switch in the implementation.

Reply:
That would mean a new RPC method for each new funding variation we think of in the future. This instead presents a set of distinct triggers similar to how we do things in the contract court.

Reply:
SGTM